Mathlib/RingTheory/DedekindDomain/Different.lean

```diff
@@ -0,0 +1,264 @@
+
+variable {A K}
+
+local notation:max I:max "ᵛ" => Submodule.traceDual A K I
+
+namespace Submodule
+
+lemma mem_traceDual {I : Submodule B L} {x} :
+    x ∈ Iᵛ ↔ ∀ a ∈ I, traceForm K L x a ∈ (algebraMap A K).range :=
+  Iff.rfl
+
+lemma le_traceDual_iff_map_le_one {I J : Submodule B L} :
+    I ≤ Jᵛ ↔ ((I * J : Submodule B L).restrictScalars A).map
+      ((trace K L).restrictScalars A) ≤ 1 := by
+  rw [Submodule.map_le_iff_le_comap, Submodule.restrictScalars_mul, Submodule.mul_le]
+  simp [SetLike.le_def, mem_traceDual]
+
+lemma le_traceDual_mul_iff {I J J' : Submodule B L} :
+    I ≤ (J * J')ᵛ ↔ I * J ≤ J'ᵛ := by
+  simp_rw [le_traceDual_iff_map_le_one, mul_assoc]
+
+lemma le_traceDual {I J : Submodule B L} :
+    I ≤ Jᵛ ↔ I * J ≤ 1ᵛ := by
+  rw [← le_traceDual_mul_iff, mul_one]
+
+lemma le_traceDual_comm {I J : Submodule B L} :
+    I ≤ Jᵛ ↔ J ≤ Iᵛ := by rw [le_traceDual, mul_comm, ← le_traceDual]
+
+lemma le_traceDual_traceDual {I : Submodule B L} :
+    I ≤ Iᵛᵛ := le_traceDual_comm.mpr le_rfl
+
+end Submodule
+
+open Submodule
+
+variable [IsIntegrallyClosed A]
+
+lemma Submodule.mem_traceDual_iff_isIntegral {I : Submodule B L} {x} :
+    x ∈ Iᵛ ↔ ∀ a ∈ I, IsIntegral A (traceForm K L x a) :=
+  forall₂_congr (fun _ _ ↦ IsIntegrallyClosed.isIntegral_iff.symm)
+
+/-- If `b` is an `A`-integral basis of `L` with discriminant `b`, then `d • a * x` is integral over
+  `A` for all `a ∈ I` and `x ∈ Iᵛ`. -/
+lemma isIntegral_discr_mul_of_mem_traceDual
+    (I : Submodule B L) {ι} [DecidableEq ι] [Fintype ι]
+    {b : Basis ι K L} (hb : ∀ i, IsIntegral A (b i))
+    {a x : L} (ha : a ∈ I) (hx : x ∈ Iᵛ) :
+    IsIntegral A ((discr K b) • a * x) := by
+  have hinv : IsUnit (traceMatrix K b).det := by
+    simpa [← discr_def] using discr_isUnit_of_basis _ b
+  have H := mulVec_cramer (traceMatrix K b) fun i => trace K L (x * a * b i)
+  have : Function.Injective (traceMatrix K b).mulVec
+  · rwa [mulVec_injective_iff_isUnit, isUnit_iff_isUnit_det]
+  rw [← traceMatrix_of_basis_mulVec, ← mulVec_smul, this.eq_iff,
+    traceMatrix_of_basis_mulVec] at H
+  rw [← b.equivFun.symm_apply_apply (_ * _), b.equivFun_symm_apply]
+  apply IsIntegral.sum
+  intro i _
+  rw [smul_mul_assoc, b.equivFun.map_smul, discr_def, mul_comm, ← H, Algebra.smul_def]
+  refine RingHom.IsIntegralElem.mul _ ?_ (hb _)
+  apply IsIntegral.algebraMap
+  rw [cramer_apply]
+  apply IsIntegral.det
+  intros j k
+  rw [updateColumn_apply]
+  split
+  · rw [mul_assoc]
+    rw [mem_traceDual_iff_isIntegral] at hx
+    apply hx
+    have ⟨y, hy⟩ := (IsIntegralClosure.isIntegral_iff (A := B)).mp (hb j)
+    rw [mul_comm, ← hy, ← Algebra.smul_def]
+    exact I.smul_mem _ (ha)
+  · exact isIntegral_trace (RingHom.IsIntegralElem.mul _ (hb j) (hb k))
+
+variable (A K)
+
+open scoped Classical
+
+/-- The dual of a non-zero fractional ideal is the dual of the submodule under the traceform. -/
+noncomputable
+def FractionalIdeal.dual (I : FractionalIdeal B⁰ L) :
```

Comment:
You can `namespace FractionalIdeal` here. Also `variable {I : FractionalIdeal B⁰ L}` is probably a good idea.